src/main/java/org/wildfly/security/sasl/digest/AbstractDigestMechanism.java

```diff
@@ -362,64 +353,49 @@ protected DigestWrapper(boolean confidential) {
-            key.append(CLIENT_MAGIC_INTEGRITY);
-        } else {
-            key.append(SERVER_MAGIC_INTEGRITY);
-        }
-        md5.reset();
-        byte[] ki = md5.digest(key.toArray());
-
-        byte[] messageMac = computeHMAC(ki, seqNum, message, offset, len);
+        byte[] messageMac = DigestUtils.computeHMAC(wrapHmacKeyIntegrity, wrapSeqNum, hmacMD5, message, offset, len);
 
         byte[] result = new byte[len + 16];
         System.arraycopy(message, offset, result, 0, len);
         System.arraycopy(messageMac, 0, result, len, 10);
-        integerByteOrdered(1, result, len + 10, 2);  // 2-byte message type number in network byte order with value 1
-        integerByteOrdered(seqNum, result, len + 12, 4); // 4-byte sequence number in network byte order
-        seqNum++;
+        DigestUtils.integerByteOrdered(1, result, len + 10, 2);  // 2-byte message type number in network byte order with value 1
+        DigestUtils.integerByteOrdered(wrapSeqNum, result, len + 12, 4); // 4-byte sequence number in network byte order
+        wrapSeqNum++;
         return result;
     }
 
     private byte[] unwrapIntegrityProtectedMessage(byte[] message, int offset, int len) throws SaslException {
 
-        int messageType = decodeByteOrderedInteger(message, offset + len - 6, 2);
-        int extractedSeqNum = decodeByteOrderedInteger(message, offset + len - 4, 4);
+        int messageType = DigestUtils.decodeByteOrderedInteger(message, offset + len - 6, 2);
+        int extractedSeqNum = DigestUtils.decodeByteOrderedInteger(message, offset + len - 4, 4);
 
         if (messageType != 1) {
             throw new SaslException("MessageType must equal to 1, but is different");
         }
 
-        ByteStringBuilder key = new ByteStringBuilder(hA1);
-        if (format == FORMAT.CLIENT) {
-            key.append(SERVER_MAGIC_INTEGRITY);
-        } else {
-            key.append(CLIENT_MAGIC_INTEGRITY);
+        if (extractedSeqNum != unwrapSeqNum) {
+            throw new SaslException("Bad sequence number while unwrapping");
         }
-        md5.reset();
-        byte[] ki = md5.digest(key.toArray());
 
         byte[] extractedMessageMac = new byte[10];
         byte[] extractedMessage = new byte[len - 16];
         System.arraycopy(message, offset, extractedMessage, 0, len - 16);
         System.arraycopy(message, offset + len - 16, extractedMessageMac, 0, 10);
 
-        byte[] expectedHmac = computeHMAC(ki, extractedSeqNum, extractedMessage, 0, extractedMessage.length);
+        byte[] expectedHmac = DigestUtils.computeHMAC(unwrapHmacKeyIntegrity, extractedSeqNum, hmacMD5, extractedMessage, 0, extractedMessage.length);
 
         // validate MAC block
         if (Arrays2.equals(expectedHmac, 0, extractedMessageMac, 0, 10) == false) {
-            throw new SaslException("MAC validation failed while unwrapping");
+            return NO_BYTES; // packet by attacker should be ignored, should not interrupt user communication
```

Comment:
Overall looks good, comments so far from me are mainly cosmetic - just this part I am not sure of.  Do we have anything in the latest RFC that mandates how we should handle this?   I see you are trying to prevent something that could become the basis for denial of service, but at the same time silently ignoring a failed mac validation does not feel correct either.